### ISSUE TITLE
[feature request] nn.Transformer

### ISSUE THREAD

**Issue by aconneau** [pytorch/pytorch#10459]
## feature request description

The NLP community is shifting from LSTMs to [Transformers](https://arxiv.org/abs/1706.03762) for a number of NLP tasks.
This would be great if we could have a packed standard Transformer implementation in the 'nn' package, i.e. a **nn.Transformer**, just like we have a nn.LSTM.

## Code example

fairseq-py has a tested implementation of the Transformer encoder and decoder (for LM and NMT) https://github.com/pytorch/fairseq/blob/master/fairseq/models/transformer.py. Wrapping this up in 'nn' should be the most straightforward approach to reimplementing this model.

Thanks!

**Comment by zou3519** [comment]
We'll look more into this. This is definitely something we could have in contrib/, the question really is should this be something we have in core.

**Comment by soumith** [comment]
it's one of the most used methods in NLP, so we should definitely have it in core.

**Comment by zhedongzheng** [comment]
Transformer is a large concept and there are a lot variants applied to different tasks in the field of NLP.
Adding layers (e.g. nn.SelfAttention, nn.sinusoidalPositionalEncoding, nn.FeedForwardLayer) maybe makes more sense.

**Comment by aconneau** [comment]
@zhedongzheng: I haven't seen many variants of it though. Apart from [that discussion](https://github.com/pytorch/fairseq/blob/master/fairseq/models/transformer.py#L362-L371) on the regularization (things we can include as parameters I guess), the tensor2tensor implementation can be seen as the default one.

**Comment by zhedongzheng** [comment]
@aconneau I agree with you now.
the default Transformer consists of encoders and decoders, which is mainly used for translation. Therefore, if we want to apply on other tasks such as text classification, we need to at least remove the decoders.
Therefore, maybe calling nn.TransformerEncoder is what you mean.

**Comment by shawnjhenry** [comment]
I've seen a few variants of the transformer architecture floating around, including one for monolingual text generation used here: https://arxiv.org/pdf/1801.10198.pdf to generate Wikipedia articles from source documents and here: https://arxiv.org/pdf/1808.02622.pdf to generate physician notes from electronic health records.  The main difference is that, rather than vanilla self attention, this model uses what the authors call "local attention" (basically just regular self attention sharded into more manageable sizes) and "memory compressed attention" (the keys and values are length compressed using a strided 1d convolution before applying the self attention).  This model would be impossible to implement with a monolithic nn.Transformer module unless it includes the sharding and memory compression options.  There may be benefits to implementing the layers that comprise the transformer architecture individually as suggested above.

**Comment by Kaixhin** [comment]
Multi-head dot product attention is also used in the [relational memory core](https://arxiv.org/abs/1806.01822) and in place of the (pairwise) relation net in ["relational deep reinforcement learning"](https://arxiv.org/abs/1806.01830), so the concept is more general than NLP. More broadly there's also [non-local blocks](https://arxiv.org/abs/1711.07971), so it may be worth considering how they might share abstract components.

Going by [The Annotated Transformer](http://nlp.seas.harvard.edu/2018/04/03/attention.html), it seems like multi-head attention is the most non-trivial module, then positional encodings, and then it's generally a composition of existing layers. In the same way that `torchvision` contains ResNets and other CV models, wouldn't a full transformer model for NLP fit better into `torchtext`?

**Comment by aconneau** [comment]
There is now a new Transformer in town, the ["universal transformer"](https://ai.googleblog.com/2018/08/moving-beyond-translation-with.html) with [its implementation here](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/research/universal_transformer.py).

**Comment by c0nn3r** [comment]
I don't think Transformers should be added for a reason demonstrated in this thread.

LSTMs are a [very mature model](http://www.bioinf.jku.at/publications/older/2604.pdf) (introduced in 1997) and it has been demonstrated that it is [very hard to improve with over a properly turning LSTM](https://arxiv.org/abs/1707.05589). However, a major improvement to the LSTM: the forget gate, [was introduced in 1999](https://pdfs.semanticscholar.org/1154/0131eae85b2e11d53df7f1360eeb6476e7f4.pdf), two years later from the original paper. At this point after many papers evaluating LSTM variants it's clear what type of API is needed for most LSTM applications and I don't think this is true for the Transformer.

The problem with adding Transformers to core is because there will always be "a new Transformer town" (at least it seems for the next few months ;) ) and I think there needs to be evaluation after the dust has settled.

Another question is tradeoffs. If the implementation in core is focused around speed then it will be hard to tweak (granted the JIT will help with this issue somewhat) and if it focused on usability then there will always be [implementations that outperform it](https://github.com/marian-nmt/marian).

**Comment by ArmenAg** [comment]
This is a good point. Transformer as an end to end model might not be mature but in NLP we are seeing a general trend toward utilization of attention. Even if we're not to implement the Transfomer I still believe it is necessary to implement the common attention modules.

In general this is a question of how agile PyTorch as a framework will be. If we're too agile we might have a mess on our hands along the lines of tf.contrib, but on the other hand only implementing the very core modules in deep learning will lead to a lot of code duplication or bugs across various projects/codebases. The debate should be where this sweet spot is.

And to play devils advocate to your point about Transformer, in the short time that they have been out they're showing [better performance of LSTMs](https://arxiv.org/abs/1808.04444). Maybe we should't go all out in writing a highly performant C++ implementation of Transformers, but a well tested standardized python version would be of great help to the NLP community. At least common attention modules should be implemented.

**Comment by c0nn3r** [comment]
Sure, but I think this more speaks to the need for better ways to [handle padding](https://github.com/salesforce/matchbox)(something that comes up a lot when implementing Transformer models). Integrating this in the JIT is a great example of this. Named dimensions are another way we can allow Transformer code immensely more readable, here is an example of this:

Before:
```python
class Attention(nn.Module):
  def __init__(self, query_dim, key_dim, value_dim):
    super(Attention, self).__init__()
    self.scale = 1. / math.sqrt(query_dim)

  def forward(self, query, keys, values):
    # Query = [BxQ]
    # Keys = [TxBxK]
    # Values = [TxBxV]
    # Outputs = a:[TxB], lin_comb:[BxV]

    # Here we assume q_dim == k_dim (dot product attention)

    query = query.unsqueeze(1) # [BxQ] -> [Bx1xQ]
    keys = keys.transpose(0,1).transpose(1,2) # [TxBxK] -> [BxKxT]
    energy = torch.bmm(query, keys) # [Bx1xQ]x[BxKxT] -> [Bx1xT]
    energy = F.softmax(energy.mul_(self.scale), dim=2) # scale, normalize

    values = values.transpose(0,1) # [TxBxV] -> [BxTxV]
    linear_combination = torch.bmm(energy, values).squeeze(1) #[Bx1xT]x[BxTxV] -> [BxV]
    return energy, linear_combination
```
These types of comments are very common on Transformer implementations to keep track of the current dimensions. 

After:
```python
class Attention(nn.Module):
  def __init__(self, query_dim, key_dim, value_dim):
    super(Attention, self).__init__()
    self.scale = 1. / math.sqrt(query_dim)

  def forward(self, query, keys, values):
    query = query.unsqueeze('batch')
    keys = keys.transpose('timesteps', 'batch').transpose('timesteps', 'keys')
    energy = torch.bmm(query, keys)
    energy = F.softmax(energy.mul_(self.scale), dim='timesteps')

    values = values.transpose('timesteps', 'batch')
    linear_combination = torch.bmm(energy, values).squeeze('timesteps')
    return energy, linear_combination
```
(Sorry if I made any mistakes converting this) Why not better enable people to write better Transformer and attention based code?

Yes, Transformers have shown impressive performance in a variate of tasks, but if anything this speaks to my point that there are potentially even larger gains to be had.

**Comment by ArmenAg** [comment]
The discussion on named dimensions should be moved to a different issue. If we are to agree that some common attention models need to be implemented we should also discuss masking\padding.

The current way I deal with masking is by passing a secondary byte tensor marking with 1 all the padding locations. From there you can do something like so:

```
class ScaledDotProductAttention(nn.Module):
    """Scaled Dot Product Attentin from the Attention is All You Need paper.
        Refs: https://arxiv.org/abs/1706.03762
    """

    def __init__(self, dropout: float=0.1):
        """
            Args:
                dropout: float ; s.e.
        """
        super(ScaledDotProductAttention, self).__init__()
        self.dropout = nn.Dropout(dropout)

    def forward(self, query: torch.Tensor, keys: torch.Tensor, values: torch.Tensor, attn_mask: Optional[torch.ByteTensor]=None) -> torch.Tensor:
        """
            Args:
                query: torch.Tensor(batch_size, time_step_query, d_model) ; query tensor
                keys: torch.Tensor(batch_size, time_step_keys, d_model) ; keys tensor
                values: torch.Tensor(batch_size, time_step_keys, d_value) ; values tensor
                attn_mask: Optional[torch.ByteTensor(batch_size, time_step_query)] ; attn mask tensor
            Returns:
                output: torch.Tensor(batch_size, time_step_query, d_value) ; output tensor
        """
        attn = torch.bmm(query, keys.transpose(1, 2))
        attn = attn.div(query.size(2)**0.5)

        if attn_mask is not None:
            attn_mask = attn_mask.unsqueeze(2).expand_as(attn)
            # instability with -inf
            attn.data.masked_fill_(attn_mask, -100)

        attn = F.softmax(attn, dim=2)
        attn = self.dropout(attn)
        output = torch.bmm(attn, values)

        return output
```

@soumith @apaszke Do you have any comments on this thread?

**Comment by Evpok** [comment]
@ArmenAg From what I gather of the current work in jit, the current focus is precisely on being able to do away with padding and such for minibatching by having automated conversion from single sample code to batch.

**Comment by mttk** [comment]
It would be really important to start populating `torch/contrib` a bit more with building blocks and complete model implementations, similar to that other large DL library. The transformer would fit under the complete model umbrella, while various attention implementations & positional encoding are building blocks.

I have a file with a bunch of implemented models which I keep copy/pasting to every new project, but I'm sure that there are a lot of inefficiencies in the implementations. For this reason it would be awesome to have a lot of people stress testing and improving an "official" implementation, and the decisions whether to add some component and where to add it to can be handled via PR's.

There's a lot of great libraries such as fairseq and AllenNLP, but I dislike importing a whole library and all of its dependencies / overhead if I just need one module. Having clean, optimized and minimal implementations would be pretty important.

**Comment by hmc-cs-mdrissi** [comment]
If you do have self attention as its own separate building block, the main thing I'd want included as an option is to properly deal with masking future time steps. It's one of the things that's easier to forget if you're implementing it yourself and unlike some other mistakes, won't hurt you during training (in fact it'll make you over estimate your performance).

**Comment by ArmenAg** [comment]
It seems like the conclusion reached is that we should implement common attention modules in core, while leaving end-to-end architectures to be cleanly implemented in contrib. As long as we're vigilant about what code ends up in contrib/ this approach makes sense.

In terms of padding do we want to use `torch.nn.utils.rnn.pack_padded_sequence()`? In my opinion it's easier to pass through a boolean tensor with 1's representing pad values in order to mask. It seems like this is how fairseq does it and how our team does it as well. I can pick up this work item once the masking api is somewhat clear.

**Comment by mfuntowicz** [comment]
FYI, CuDNN 7.5 brings some dedicated APIs for Multi-Head Attentions as commonly used in Transformer-based layers.

> A new set of APIs are added to provide support for Multi-Head Attention computation. 

I would be very interesting to provide, if not a Transformer Layer, a Multihead Attention mechanism leveraging such APIs on GPU.

CuDNN ref : https://docs.nvidia.com/deeplearning/sdk/cudnn-release-notes/rel_750.html#rel_750

**Comment by Squadrick** [comment]
Can I work on this issue?

**Comment by mttk** [comment]
I believe it is somewhat resolved: 
MHA -- https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/activation.py#L682
Transformer -- https://github.com/pytorch/pytorch/pull/20170/files

**Comment by zou3519** [comment]
cc @zhangguanheng66 who has a PR for nn.Transformer

**Comment by zhangguanheng66** [comment]
@Squadrick Thanks for the offer. We will land the PR very soon.

**Comment by calclavia** [comment]
Does the current MHA implementation use cudnn? Could lead to faster computation as @mfuntowicz suggested.

**Comment by zhangguanheng66** [comment]
@calclavia to your question, no. We are still waiting to bind the cudnn implementation of MHA. Once the cudnn implementation is more mature, we will move forward.

**Comment by hadaev8** [comment]
@zhangguanheng66 
Any update about cudnn MHA?

**Comment by zhangguanheng66** [comment]
Last time, I saw some performance benchmark on cudnn MHA and pytorch MHA. We are working on decomposing MHA to add more flexibility.
